File: lib/pylib/dataset.py
from config import *
import os, json, math

class DatasetView:
	dataset = None
	x = 0
	y = 0
	files_group = None
		
	def __init__(self, dataset_, x_, y_, files_group_=None):
		self.dataset = dataset_
		self.x = x_
		self.y = y_
		self.files_group = files_group_

	def local_parameters(self):
		if self.files_group is None: return self.dataset.parameters
		else: return self.dataset.parameters[self.files_group]

	def local_x(self):
		par = self.local_parameters()
		x = self.x
		if "x_index_factor" in par: x = math.trunc(x * par["x_index_factor"])
		if "x_index_offset" in par: x = x + par["x_index_offset"]
		return x

	def local_y(self):
		par = self.local_parameters()
		y = self.y
		if "y_index_factor" in par: y = math.trunc(y * par["y_index_factor"])
		if "y_index_offset" in par: y = y + par["y_index_offset"]
		return y
		
	def local_view(self, name):
		if name not in self.dataset.parameters: raise Exception("no {} parameters".format(name))
		return DatasetView(self.dataset, self.x, self.y, name)

	def format_filename(self, tpl):
		if self.dataset.is_2d(): relpath = tpl.format(x=self.local_x(), y=self.local_y())
		else: relpath = tpl.format(x=self.local_x())
		return self.dataset.filepath(relpath)
		
	def format_name(self, tpl):
		if self.dataset.is_2d(): return tpl.format(x=self.x, y=self.y)
		else: return tpl.format(x=self.x)
			
	def camera_name(self):
		return self.format_name(self.dataset.parameters["camera_name_format"])

	def image_filename(self):
		return self.format_filename(self.dataset.parameters["image_filename_format"])
	def depth_filename(self):
		return self.format_filename(self.dataset.parameters["depth_filename_format"])
		
	def vsrs(self):
		return self.local_view("vsrs")
	def kinect_raw(self):
		return self.local_view("kinect_raw")


class Dataset:
	parameters = None
	dirname = None
	x_index_range = None
	y_index_range = None
	
	def __init__(self, parameters_filename):
		self.dirname = os.path.dirname(parameters_filename)
		with open(parameters_filename) as f:
			self.parameters = json.load(f)
		
		self.x_index_range = self.parameters["x_index_range"]
		if "y_index_range" in self.parameters:
			self.y_index_range = self.parameters["y_index_range"]		

	def filepath(self, relpath):
		return os.path.join(self.dirname, relpath)		

	def cameras_filename():
		return self.filepath(self.parameters["cameras_filename"])

	def is_2d(self):
		return (self.y_index_range is not None)
	def is_1d(self):
		return not self.is_2d()

	def x_min(self):
		return int(self.x_index_range[0])
	def x_max(self):
		return int(self.x_index_range[1])
	def x_step(self):
		if len(self.x_index_range) == 3: return self.x_index_range[2]
		else: return 1
	def x_valid(self, x):
		return (x >= self.x_min()) and (x <= self.x_max()) and ((x - self.x_min()) % self.x_step() == 0)
	def x_count(self):
		return (self.x_max() - self.x_min() + 1) // self.x_step()		
	def x_indices(self):
		x = self.x_min()
		while x <= self.x_max():
			yield x
			x = x + self.x_step()
	
	def y_min(self):
		if self.is_2d(): return int(self.y_index_range[0])
		else: return 0
	def y_max(self):
		if self.is_2d(): return int(self.y_index_range[1])
		else: return 0
	def y_step(self):
		if self.is_2d() and len(self.y_index_range) == 3: return self.y_index_range[2]
		else: return 1
	def y_valid(self, y):
		return (y >= self.y_min()) and (y <= self.y_max()) and ((y - self.y_min()) % self.y_step() == 0)
	def y_count(self):
		return (self.y_max() - self.y_min() + 1) // self.y_step()		
	def y_indices(self):
		y = self.x_min()
		while y <= self.y_max():
			yield y
			y = y + self.y_step()

	def view(self, x, y=None):
		if y is None:
			if self.is_2d(): raise Exception("must specify y view index, for 2d dataset")
			y = 0
		if not self.x_valid(x): raise Exception("x view index out of range")
		if not self.y_valid(y): raise Exception("y view index out of range")
		return DatasetView(self, x, y)
		

File: src/lib/dataset.cc
#include "dataset.h"
#include "string.h"
#include <format.h>
#include <stdexcept>
#include <fstream>
#include <string>
#include <ostream>

namespace tlz {


int dataset_view::local_filename_x_() const {
	float factor = get_or(local_parameters(), "filename_x_index_factor", 1.0);
	int offset = get_or(local_parameters(), "filename_x_index_offset", 0);
	
	int loc_x = x_;
	loc_x *= factor;
	loc_x += offset;
	return loc_x;
}

int dataset_view::local_filename_y_() const {
	float factor = get_or(local_parameters(), "filename_y_index_factor", 1.0);
	int offset = get_or(local_parameters(), "filename_y_index_offset", 0);
	
	int loc_y = y_;
	loc_y *= factor;
	loc_y += offset;
	return loc_y;
}

std::string dataset_view::format_name(const std::string& tpl) const {
	if(dataset_.is_2d()) return fmt::format(tpl, fmt::arg("x", x_), fmt::arg("y", y_));
	else return fmt::format(tpl, fmt::arg("x", x_));
}

std::string dataset_view::format_filename(const std::string& tpl) const {
	std::string relpath;
	if(dataset_.is_2d()) relpath = fmt::format(tpl, fmt::arg("x", local_filename_x_()), fmt::arg("y", local_filename_y_()));
	else relpath =  fmt::format(tpl, fmt::arg("x", local_filename_x_()));
	return dataset_.filepath(relpath);
}

dataset_view::dataset_view(const dataset& datas, int x, int y, const std::string& grp) :
	dataset_(datas), x_(x), y_(y), group_(grp) { }


const json& dataset_view::local_parameters() const {
	if(group_.empty()) return dataset_.parameters();
	else return dataset_.parameters()[group_];
}
 
std::string dataset_view::local_filename(const std::string& name, const std::string& def) const {
	std::string tpl = get_or(local_parameters(), name, std::string());
	if(! tpl.empty()) return format_filename(tpl);
	else return def;
}

std::string dataset_view::camera_name() const {	
	return format_name(dataset_.parameters()["camera_name_format"]);
}

std::string dataset_view::image_filename() const {
	return local_filename("image_filename_format");
}

std::string dataset_view::depth_filename() const {
	return local_filename("depth_filename_format");
}

std::string dataset_view::mask_filename() const {
	return local_filename("mask_filename_format");
}

std::string dataset_view::group() const {
	return group_;
}

dataset_view dataset_view::group_view(const std::string& grp) const {
	if(grp.empty()) return dataset_view(dataset_, x_, y_);
	else if(has(dataset_.parameters(), grp)) return dataset_view(dataset_, x_, y_, grp);
	else throw std::runtime_error("no group '" + grp + "' in dataset");
}

/////


dataset::dataset(const std::string& parameters_filename) {
	std::size_t last_sep_pos = parameters_filename.find_last_of('/');
	if(last_sep_pos == std::string::npos) dirname_ = "./";
	else dirname_ = parameters_filename.substr(0, last_sep_pos + 1);
	
	std::ifstream stream(parameters_filename);
	stream >> parameters_;

	for(int v : parameters_["x_index_range"]) x_index_range_.push_back(v);
	if(parameters_.count("y_index_range") == 1)
		for(int v : parameters_["y_index_range"]) y_index_range_.push_back(v);
}

bool dataset::is_1d() const {
	return (y_index_range_.size() == 0);
}

bool dataset::is_2d() const {
	return (y_index_range_.size() > 0);
}

const json& dataset::group_parameters(const std::string& grp) {
	if(grp.empty()) return parameters_;
	else if(has(parameters_, grp)) return parameters_[grp];
	else throw std::runtime_error("no group '" + grp + "' in dataset");
}

std::string dataset::filepath(const std::string& relpath) const {
	return dirname_ + relpath;
}

std::string dataset::cameras_filename() const {
	return filepath(parameters_["cameras_filename"]);
}

int dataset::x_min() const {
	return x_index_range_[0];
}

int dataset::x_max() const {
	return x_index_range_[1];
}

int dataset::x_step() const {
	if(x_index_range_.size() == 3) return x_index_range_[2];
	else return 1;
}

bool dataset::x_valid(int x) const {
	return (x >= x_min()) && (x <= x_max()) && (((x - x_min()) % x_step()) == 0);
}

int dataset::x_count() const {
	return (x_max() - x_min() + x_step()) / x_step();
}

int dataset::x_mid() const {	
	return x_min() + (((x_max() - x_min()) / (2 * x_step())) * x_step());
}

std::vector<int> dataset::x_indices() const {
	std::vector<int> indices;
	for(int x = x_min(); x <= x_max(); x += x_step()) indices.push_back(x);
	return indices;
}

int dataset::y_min() const {
	if(is_2d()) return y_index_range_[0];
	else return 0;
}

int dataset::y_max() const {
	if(is_2d()) return y_index_range_[1];
	else return 0;
}

int dataset::y_step() const {
	if(is_2d() && (y_index_range_.size() == 3)) return y_index_range_[2];
	else return 1;
}

bool dataset::y_valid(int y) const {
	return (y >= y_min()) && (y <= y_max()) && (((y - y_min()) % y_step()) == 0);
}

std::vector<int> dataset::y_indices() const {
	std::vector<int> indices;
	for(int y = y_min(); y <= y_max(); y += y_step()) indices.push_back(y);
	return indices;
}

int dataset::y_count() const {
	if(is_2d()) return (y_max() - y_min() + y_step()) / y_step();
	else return 1;
}

int dataset::y_mid() const {
	if(is_2d()) return y_min() + (((y_max() - y_min()) / (2 * y_step())) * y_step());
	else return 0;
}

dataset_view dataset::view(int x) const {
	if(is_2d()) throw std::runtime_error("must specify y view index, for 2d dataset");
	if(! x_valid(x)) throw std::runtime_error("x view index out of range");
	return dataset_view(*this, x, 0);
}

dataset_view dataset::view(int x, int y) const {
	if(! x_valid(x)) throw std::runtime_error("x view index out of range");
	if(! y_valid(y)) throw std::runtime_error("y view index out of range");
	return dataset_view(*this, x, y);
}

dataset_view dataset::view(view_index idx) const {
	if(is_2d()) {
		if(! idx.is_2d()) throw std::runtime_error("must specify 2d view index");
		return view(idx.x, idx.y);
	} else {
		if(idx.is_2d()) throw std::runtime_error("must specify 1d view index");
		return view(idx.x);
	}
}


//////////


std::string encode_view_index(view_index idx) {
	std::string key = std::to_string(idx.x);
	if(idx.is_2d()) key += "," + std::to_string(idx.y);
	return key;
}

view_index decode_view_index(const std::string& key) {
	view_index idx;
	auto j_idx = explode_from_string<int>(',', key);
	idx.x = j_idx[0];
	if(j_idx.size() == 2)  idx.y = j_idx[1];
	else idx.y = -1;
	return idx;
}

std::ostream& operator<<(std::ostream& stream, const view_index& idx) {
	if(idx.is_2d()) stream << '(' << idx.x << ", " << idx.y << ')';
	else stream << '(' << idx.x << ')';
	return stream;
}


}
